packages/swingset-liveslots/src/collectionManager.js

```diff
@@ -432,6 +432,9 @@ export function makeCollectionManager(
           }
         }
       }
+      if (passStyleOf(key) === 'remotable' && getOrdinal(key) === undefined) {
```

Comment:
Hm, this introduces an extra `syscall.vatstoreGet` for  each `store.set`: `getOrdinal` does one, and then `keyToDBKey` does it again (via `encodeKey` -> `makeEncodePassable` -> `encodeRemotable` -> `getOrdinal`).

I managed to make our `has` method avoid it:

```js
      if (passStyleOf(key) === 'remotable') {
        return getOrdinal(key) !== undefined;
      } else {
        return syscall.vatstoreGet(keyToDBKey(key)) !== undefined;
      }
```

but in `get`, the extra `getOrdinal()` we added (to get a unified error message) causes the same extra `vatstoreGet` syscall in the successful case:

```js
    function get(key) {
      const { keyShape, label } = getSchema();
      mustMatch(key, keyShape, makeInvalidKeyTypeMsg(label));
      if (passStyleOf(key) === 'remotable' && getOrdinal(key) === undefined) {
        throw Fail`key ${key} not found in collection ${q(label)}`;
      }
      const result = syscall.vatstoreGet(keyToDBKey(key));
      if (!result) {
        throw Fail`key ${key} not found in collection ${q(label)}`;
      }
      return unserializeValue(JSON.parse(result));
    }
```

So, the extra syscall is unfortunate, but no worse than what `get` is already doing, and `get` probably happens more frequently than `set`.

Let's keep the perf impact in the back of our mind, and see if we can find a sensible refactoring that will let us avoid it later. If this were Python, I'd have `encodeRemotable` throw a recognizable exception type, and have `get`/`set` catch it and re-throw the same `IndexError` it would throw when `result` is missing. But I know we don't use JS error values that way.

Reply:
I guess the existing inefficiency is in `get` (`getOrdinal` → **vatstoreGet**, `keyToDBKey` → `encodeKey` → `encodeRemotable` → `getOrdinal` → **vatstoreGet**, **vatstoreGet**), which this PR moves into `mustGet` and thereby infects `set`. I agree with you about the Pythonic approach, and avoiding that here would incline me towards updating `keyToDBKey` to distinguish top-level remotables and eliminate the redundant `getOrdinal` calls:
```diff
+    const encodeOrdinal = (ordinal, remotable) =>
+      `r${zeroPad(ordinal, BIGINT_TAG_LEN)}:${convertValToSlot(remotable)}`;
+
     const encodeRemotable = remotable => {
       // eslint-disable-next-line no-use-before-define
       const ordinal = getOrdinal(remotable);
       ordinal !== undefined || Fail`no ordinal for ${remotable}`;
-      const ordinalTag = zeroPad(ordinal, BIGINT_TAG_LEN);
-      return `r${ordinalTag}:${convertValToSlot(remotable)}`;
+      return encodeOrdinal(ordinal, remotable);
     };
```
```diff
-    function keyToDBKey(key) {
-      const encodedKey = encodeKey(key);
+    function keyToDBKey(key, label = undefined) {
+      let encodedKey;
+      if (passStyleOf(key) === 'remotable') {
+        const ordinal = getOrdinal(key) || failNotFound(key, label);
+        encodedKey = encodeOrdinal(ordinal, key);
+      } else {
+        encodedKey = encodeKey(key);
+      }
       assert(encodedKey.length < MAX_DBKEY_LENGTH, 'key too large');
       return prefix(encodedKey);
     }
```
```diff
     function mustGet(key, label) {
-      if (passStyleOf(key) === 'remotable' && getOrdinal(key) === undefined) {
-        failNotFound(key, label);
-      }
-      const dbKey = keyToDBKey(key);
+      const dbKey = keyToDBKey(key, label);
       const result = syscall.vatstoreGet(dbKey) || failNotFound(key, label);
       return { dbKey, result };
     }
```
Should I expand this PR to include such changes?

Reply:
> ... If this were Python, I'd have `encodeRemotable` throw a recognizable exception type, and have `get`/`set` catch it and re-throw the same `IndexError` it would throw when `result` is missing. But I know we don't use JS error values that way.

We seem to use JS non-error values that way:

> For non-local non-erroneous control flow, our recommended practice (which we've only recently decided on and started practicing) is that **normal non-erroneous control flow**, not indicating a problem that should be brought to human attention, **should happen by raising a non-error**. -- [Errors and Control Flow](https://github.com/Agoric/agoric-sdk/wiki/Errors-and-Control-Flow)

Reply:
p.s. here's hoping for ejectors...

 - https://github.com/endojs/endo/issues/414